src/librustc_mir/borrow_check/region_infer/mod.rs

```diff
@@ -67,6 +67,12 @@ pub struct RegionInferenceContext<'tcx> {
     /// compute the values of each region.
     constraint_sccs: Rc<Sccs<RegionVid, ConstraintSccIndex>>,
 
+    /// SCCs in "dependency order" (or "post order"), meaning that if S1 -> S2,
+    /// then S2 appears first. If you process the SCCs in this order, then you
+    /// are always ensured that when you proces a given SCC, all of its
```

Comment:
```suggestion
    /// are always ensured that when you process a given SCC, all of its
```

Reply:
(I haven't read this module in a while, or this PR in detail, so take this with a grain of salt -- it seems likely a lot of this commit could be simplified: Tarjan's algorithm already discovers SCCs in postorder, and it looks like you could compute a lot of the data, a TC of the SCC values, computing the representatives, while you do the single walk)

Reply:
Hmm, that's probably true.

Reply:
IIRC some of Nuutila's works describe straightforward Tarjan modifications to handle similar problems.

Reply:
I'm not inclined to do the larger potential refactorings here, but I can check whether we can trivially ensure that the `all_sccs` is always in dependency order or something like that.

Reply:
Ok then I'll try to apply this idea here in the future